aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,670 @@
+    }
+
+    /// Takes `amount` of `coin::Coin<CoinType>` coins from `sender`, wraps them inside a `VeiledCoin<CoinType>` and\
+    /// sends to `recipient`.
+    public entry fun wrap_to<CoinType>(sender: &signer, recipient: address, amount: u64) acquires VeiledCoinMinter, VeiledCoinStore {
+        let c = coin::withdraw<CoinType>(sender, amount);
+
+        let vc = wrap_from_coin(c);
+
+        deposit<CoinType>(recipient, vc)
+    }
+
+    /// Takes `amount` of `VeiledCoin<CoinType>` coins from `sender`, unwraps them to a coin::Coin<CoinType>,
+    /// and sends them back to `sender`.`amount_ct` must be an el-gamal ciphertext encrypting `amount`
+    /// with no randomness. It is not possible for this function to be made private. Requires a range
+    /// proof on the new balance of the sender, to ensure the sender has enough money to send, in addition to a 
+    /// range proof on the transferred amount.  
+    public entry fun unwrap<CoinType>(
+	sender: &signer, 
+	amount: u64, 
+	amount_ct_bytes: vector<u8>, 
+	range_proof_updated_balance: vector<u8>, 
+	range_proof_transferred_amount: vector<u8>) acquires VeiledCoinStore, VeiledCoinMinter
+    {
+	unwrap_to<CoinType>(sender, signer::address_of(sender), amount, amount_ct_bytes, range_proof_updated_balance, range_proof_transferred_amount)
+    }
+
+    /// Takes `amount` of `VeiledCoin<CoinType>` coins from `sender`, unwraps them to a coin::Coin<CoinType>,
+    /// and sends them to `recipient`.`amount_ct` must be an el-gamal ciphertext encrypting `amount`
+    /// with no randomness. It is not possible for this function to be made private. Requires a range
+    /// proof on the new balance of the sender, to ensure the sender has enough money to send, in addition to a 
+    /// range proof on the transferred amount.  
+    public entry fun unwrap_to<CoinType>(
+	sender: &signer, 
+	recipient: address, 
+	amount: u64, 
+	amount_ct_bytes: vector<u8>, 
+	range_proof_updated_balance_bytes: vector<u8>, 
+	range_proof_transferred_amount_bytes: vector<u8>) acquires VeiledCoinStore, VeiledCoinMinter
+    {
+	let amount_ct = elgamal::new_ciphertext_from_bytes(amount_ct_bytes);
+	assert!(std::option::is_some(&amount_ct), EDESERIALIZATION_FAILED);	
+
+	let range_proof_updated_balance = bulletproofs::range_proof_from_bytes(range_proof_updated_balance_bytes);
+	let range_proof_transferred_amount = bulletproofs::range_proof_from_bytes(range_proof_transferred_amount_bytes);
+	
+	let c = unwrap_to_coin<CoinType>(sender, std::option::extract(&mut amount_ct), amount, &range_proof_updated_balance, &range_proof_transferred_amount);
+	coin::deposit<CoinType>(recipient, c);
+    }
+
+    /// Sends the specified private amount to `recipient` and updates the private balance of `sender`. Requires a range
+    /// proof on the new balance of the sender, to ensure the sender has enough money to send, in addition to a 
+    /// range proof on the transferred amount. Also requires a sigma protocol to prove that 
```

Comment:
So we need to find away to avoid repeating comments about the range proof and the sigma protocol proofs required in each operation.

It seems like we have various kinds of proofs (e.g., sigma-and-2*range, no-sigma-and-2*range, etc), which we'll need different names/structs for these. And we can explain what they do above the struct.

Reply:
So after updating unwrap to not use the second range proof we only have sigma-and-2range and no-sigma-and-1range. I think we only need a struct for the former

Reply:
Okay, let's see how that looks then!